README.md
# fetch-rewards Exercise

Program to check the health of a set of HTTP endpoints

<br>


<br>

# Table of Contents
1. [Getting Started](#getting-started)
    * [Using Docker](#running-docker)
    * [Using python locally](#running-locally)
2. [Time Complexity](time-complexity)
3. [Space Complexity](space-complexity)
2. [Scenarios to deploy](#scenarios-to-deploy)
    * [AWS Deployment Scenario](#aws-deployment-scenario)
    * [Notification requirements](#notification-requirements)

<br>

<a name="get"></a>

## Getting Started

These instructions will guide you on how to run the Python script in a Docker container.

<a name="usage"></a>

###   1. Add input file 
    Add required input file containing all the urls in the main directory as "input.yaml"

###   2. Build the Docker Image

&nbsp; Navigate to the project directory and run the following command to build the Docker image:


    docker build -t fetch .

###   3. Run the Docker Container or run locally
Once the image is built, run the Docker container:

### Running Docker
```bash
docker run fetch
```
### Running locally

```bash
python app.py
```

<br>



### Time Complexity:

1. Reading YAML file and processing entries in main functions:

    The time complexity for reading the YAML file and processing each entry is O(n), where n is the number of entries in the YAML file.

2. HTTP requests in calculate_latency function:

    The time complexity of making an HTTP request is generally O(1) for simple requests. However, the actual time may vary depending on the response time from the server.

3. Updating url_dict in main function:

    The time complexity of updating the url_dict using Counter.update is O(k), where k is the number of latencies for a specific URL in a single entry.

4. Printing availability in print_availability function:

    The time complexity for printing the availability is O(m), where m is the number of unique URLs.

5. Overall, the time complexity is dominated by the number of entries in the YAML file and the number of unique URLs.

<br>

### Space Complexity:
1. url_dict dictionary:

    The space complexity of the url_dict is O(m * k), where m is the number of unique URLs, and k is the average number of latencies for a specific URL.

2. Local variables and data structures:

    The space complexity of other local variables and data structures is O(1) as they don't depend on the input size.


<br>


## Scenarios to deploy  


### AWS Deployment Scenario
When deploying a health check service to AWS with SNS (Simple Notification Service) for notifications, there are several considerations and steps to follow. Here's a guide with information on how to set up such a deployment:

1. Architecture Overview:
AWS ECS (Elastic Container Service) can be used to deploy the health check service in a Docker container.
Implement a health check mechanism within the service to periodically check the URLs.
Integrate AWS SNS for sending notifications when health check failures occur.
2. AWS Resources:
ECS Cluster:
Create an ECS cluster to manage the deployment of containers.
Task Definition:
Define an ECS task definition specifying the Docker image, resource requirements, and environment variables.
Service:
Create an ECS service to run and maintain a specified number of task instances.
To deploy in Kubernetes use EKS and run it as sidecar to the main application
3. Integration with AWS SNS:
Create SNS Topic:
Create an SNS topic to which notifications will be published.
Subscribe Endpoints:
Subscribe relevant endpoints (email addresses, SMS numbers, etc.) to the SNS topic.
4. Service Configuration:
Environment Variables:
Set environment variables in the ECS task definition for configuring the health check service.
SNS_TOPIC_ARN: ARN of the SNS topic.
AWS_REGION: AWS region for the SNS topic.
Other environment-specific variables.
5. Health Check Service Implementation:
Monitoring URLs:
Implement logic within the health check service to periodically monitor the health of specified URLs.
SNS Notification Logic:
When a health check fails, publish a notification to the configured SNS topic.
Include relevant details such as the affected URL, error messages, and timestamps.
6. Notification Content:
Email Notifications:
For email notifications, include details in the email body or subject line.
SMS Notifications:
Keep SMS notifications concise due to character limitations.
Payload Format:
Use a standardized payload format for all notification types.
For Slack notification:
For Slack notifications, include details similar to email
7. Monitoring and Logging:
CloudWatch Metrics:
Set up CloudWatch metrics to monitor ECS service metrics, such as CPU utilization and task count.
Logging:
Implement logging within the health check service for detailed diagnostic information.
8. Security Considerations:
IAM Roles:
Create IAM roles with the necessary permissions for ECS tasks to publish to the SNS topic.
Secure Configuration:
Avoid hardcoding sensitive information in the code. Use AWS Secrets Manager or Parameter Store for secure configuration storage.
9. Testing:
End-to-End Testing:
Perform end-to-end testing to ensure that health checks trigger notifications correctly.
Simulate Failures:
Simulate URL failures to verify that notifications are sent as expected.
10. Documentation:
Deployment Documentation:
Maintain comprehensive documentation for deploying the health check service and configuring SNS notifications.
Troubleshooting Guide:
Include a troubleshooting guide for common issues and resolutions.
11. Cost Management:
Optimize Resource Usage:
Regularly review and optimize ECS resource usage to manage costs efficiently.
12. Alerting Policies:
CloudWatch Alarms:
Set up CloudWatch alarms for key ECS metrics to proactively alert on potential issues.
13. Scaling Considerations:
ECS Auto Scaling:
Consider implementing ECS Auto Scaling based on metrics like CPU or memory utilization.
14. Backup and Recovery:
ECS State Persistence:
Ensure data persistence and implement backup and recovery mechanisms if the health check service has stateful components.
15. Continuous Improvement:
Feedback Loop:
Establish a feedback loop to continuously improve the health check service based on incidents and feedback.
![Architecture](./files/Screenshot%202023-12-03%20at%2012.07.27%20PM.png)
<br>

### Notification requirements

Notifying developers when health checks are down is crucial for maintaining the reliability and availability of a system. Here are some considerations and information to include in notifications:

1. Notification Content:
a. Error Details:
Provide information about the error or issue encountered during the health check.
Include error codes, error messages, or a brief description of the problem.
b. Affected URL:
Clearly mention the URL or endpoint that is experiencing issues.
Include the environment (e.g., production, staging) to help identify the context.
c. Timestamp:
Include the timestamp of when the health check failed.
Use a standardized timestamp format for consistency.
d. Duration of Downtime:
Indicate how long the health check has been failing.
Include a duration that helps developers understand the impact.
2. Communication Channels:
a. Email:
Send detailed emails with the information mentioned above.
Use a subject line that clearly indicates a health check failure.
b. Slack or Chat Integration:
Integrate with communication channels like Slack for real-time alerts.
Post notifications to a dedicated channel for infrastructure or health monitoring.
c. SMS Alerts:
Consider sending critical alerts via SMS for immediate attention.
Use SMS selectively for high-priority issues to avoid notification fatigue.
3. Alert Severity Levels:
a. Critical Alerts:
Use critical alerts for severe issues affecting the entire system or critical functionalities.
Require immediate attention and intervention.
b. Warning Alerts:
Send warning alerts for issues that might impact certain functionalities but don't pose an immediate threat.
Allow developers to investigate and address during regular working hours.
4. Instructions for Developers:
a. Investigation Steps:
Provide steps or guidance on how developers can investigate the issue further.
Include relevant log locations, debugging tools, or monitoring dashboards.
b. Resolution Steps:
If known, suggest steps developers can take to resolve the issue.
Include links to documentation or runbooks for troubleshooting.
5. Escalation Policy:
a. Define Roles and Responsibilities:
Clearly define who is responsible for addressing health check failures.
Specify escalation paths if the primary contact is unavailable.
b. On-Call Rotation:
Implement an on-call rotation for developers to ensure 24/7 coverage.
Include contact details for the on-call developer.
6. Integration with Monitoring Tools:
    a. Link to Monitoring Dashboards:
        Include links to relevant monitoring dashboards or tools.
        Developers can quickly access additional information for troubleshooting.
7. Continuous Improvement:
    a. Post-Incident Review:
        Conduct post-incident reviews to analyze the root cause of health check failures.
        Use findings to improve the monitoring system and prevent future issues.


Example Notification:

```
    Subject: [CRITICAL] Health Check Failure - Production Environment

    Dear [Team/Developer Name],

    The health check for the URL [Affected URL] in the production environment has failed.

    Error Details: [Provide detailed error information]
    Timestamp: [Include timestamp in UTC]
    Duration of Downtime: [Specify duration]
    Action Required:

    Investigation Steps: [Provide steps for investigation]
    Resolution Steps: [If known, suggest steps for resolution]
    Monitoring Dashboard: [Include a link to the monitoring dashboard]

    For immediate attention, please contact the on-call developer [On-Call Developer Name] at [Contact Details].

    Thank you,
    [Your Team or System Name]
```

<br>

## Contributing

Pull requests are welcome. For major changes, please open an issue first
to discuss what you would like to change.

Please make sure to update tests as appropriate.

CHANGELOG.md
All notable changes will be documented in this file.


main_py_test.py
import pytest
import time
from collections import defaultdict, Counter
from unittest.mock import patch, Mock
from app import calculate_latency, process_entries, print_availability

@pytest.fixture
def mock_requests_get(mocker):
    return mocker.patch('app.requests.get')

@pytest.fixture
def mock_requests_post(mocker):
    return mocker.patch('app.requests.post')

def test_calculate_latency_get_success(mock_requests_get):
    mock_response = Mock(status_code=200)
    mock_response.elapsed.total_seconds.return_value = 0.5
    mock_requests_get.return_value = mock_response

    latency = calculate_latency("https://example.com", method="GET")
    assert latency <= 0.5

def test_calculate_latency_get_failure(mock_requests_get):
    mock_response = Mock(status_code=404)
    mock_requests_get.return_value = mock_response

    latency = calculate_latency("https://example.com", method="GET")
    assert latency is None

def test_calculate_latency_post_success(mock_requests_post):
    mock_response = Mock(status_code=200)
    mock_response.elapsed.total_seconds.return_value = 0.7
    mock_requests_post.return_value = mock_response

    latency = calculate_latency("https://example.com", method="POST", headers={}, body={})
    assert latency <= 0.7

def test_calculate_latency_post_failure(mock_requests_post):
    mock_response = Mock(status_code=404)
    mock_requests_post.return_value = mock_response

    latency = calculate_latency("https://example.com", method="POST", headers={}, body={})
    assert latency is None

def test_print_availability(capsys):
    url_dict = defaultdict(Counter)
    url_dict["Example GET"].update([0.1, 0.2, 0.6])
    url_dict["Example POST"].update([0.3, 0.4, 0.5])

    print_availability(url_dict)
    captured = capsys.readouterr()

    assert "Example GET has 66.67% availability" in captured.out
    assert "Example POST has 100.00% availability" in captured.out


main_unit_test.py
import time
from collections import defaultdict, Counter
from app import calculate_latency, process_entries, print_availability

def test_calculate_latency():
    # Example test for calculate_latency function
    url = "http://example.com"
    method = "GET"
    headers = {"User-Agent": "TestAgent"}
    body = {"key": "value"}

    latency = calculate_latency(url, method, headers, body)
    print(latency)
    assert latency is not None
    assert latency >= 0

def test_process_entries():
    # Example test for process_entries function
    file_path = "input.yaml"  # Create a test YAML file for testing
    url_dict = defaultdict(Counter)

    # Call the function
    result_url_dict = process_entries(file_path, url_dict)

    assert result_url_dict is not None
    assert isinstance(result_url_dict, defaultdict)


test_calculate_latency()
test_process_entries()

app.py
import requests
import time
import yaml
from collections import defaultdict, Counter

def calculate_latency(url, method="GET", headers=None, body=None):
    """
    Calculate latency for an HTTP request.

    Args:
        url (str): The URL for the HTTP request.
        method (str): The HTTP method (GET or POST).
        headers (dict): Headers for the request.
        body (dict): JSON body for POST requests.

    Returns:
        float or None: Latency of the request or None if unsuccessful.
    """
    start_time = time.time()
    try:
        if method.upper() == "GET" or method is None:
            response = requests.get(url)
        elif method.upper() == "POST":
            response = requests.post(url, headers=headers, json=body)
        else:
            raise ValueError("Invalid HTTP method")

        latency = time.time() - start_time
        return latency if response.status_code == 200 else None

    except requests.RequestException:
        return None

def process_entries(file_path, url_dict):
    """
    Process entries from a YAML file and update the url_dict with latencies.

    Args:
        file_path (str): Path to the YAML file.
        url_dict (defaultdict): Dictionary to store latencies for each URL.

    Returns:
        defaultdict: Updated url_dict.
    """
    with open(file_path, 'r') as file:
        data = yaml.safe_load(file)
        for entry in data:
            url = entry.get('url')
            headers = entry.get('headers', {})
            method = entry.get('method', 'GET')
            name = str(entry.get('name', '')).split()[0]
            body = entry.get('body')

            latency = calculate_latency(url, method, headers, body)
            url_dict[name].update([latency])

    return url_dict

def print_availability(url_dict):
    """
    Print the availability percentage for each URL.

    Args:
        url_dict (defaultdict): Dictionary containing latencies for each URL.
    """
    for name, latencies in url_dict.items():
        total_requests = len(latencies)
        successful_requests = sum(1 for latency in latencies if latency is not None and latency <= 0.5)
        availability_percentage = (successful_requests / total_requests) * 100 if total_requests > 0 else 0
        print(f"{name} has {availability_percentage:.2f}% availability")

if __name__ == "__main__":
    file_path = "input.yaml"

    # Using defaultdict and Counter for efficient counting
    url_dict = defaultdict(Counter)

    while True:
        url_dict = process_entries(file_path, url_dict)
        print_availability(url_dict)
        time.sleep(15)
